kv/dist_sender.go

```diff
@@ -491,154 +606,63 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
 
-				// If retryable, allow retry. For range not found or range
-				// key mismatch errors, we don't backoff on the retry,
-				// but reset the backoff loop so we can retry immediately.
-				switch err.(type) {
-				case *proto.RangeNotFoundError, *proto.RangeKeyMismatchError:
-					// Range descriptor might be out of date - evict it.
-					ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
-					// On addressing errors, don't backoff; retry immediately.
-					return retry.Reset, nil
-				case *proto.NotLeaderError:
-					leader := err.(*proto.NotLeaderError).GetLeader()
-					if leader != nil {
-						// First, verify that leader is a known replica; if not,
-						// we've got a stale replica; evict cache and store leader.
-						if i := replicas.FindReplica(leader.StoreID); i == -1 {
-							ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
-						}
-						ds.updateLeaderCache(proto.RaftID(desc.RaftID), *leader)
-					}
-					return retry.Reset, nil
-				default:
-					if retryErr, ok := err.(util.Retryable); ok && retryErr.CanRetry() {
-						return retry.Continue, nil
-					}
-				}
-				return retry.Break, err
-			}
+	curReply := finalReply
+	for {
+		call.Reply = curReply
+		curReply.Header().Reset()
 
-			// If this request has a bound, such as MaxResults in
-			// ScanRequest, check whether enough rows have been retrieved.
-			if args, ok := args.(proto.Bounded); ok && args.GetBound() > 0 {
-				bound := args.GetBound()
-				if reply, ok := reply.(proto.Countable); ok {
-					if nextBound := bound - reply.Count(); nextBound > 0 {
-						// Update bound for the next round.
-						args.SetBound(nextBound)
-					} else {
-						// Set flag to break the loop.
-						descNext = nil
-					}
-				}
+		var desc, descNext *proto.RangeDescriptor
+		err := retry.WithBackoff(retryOpts, func() (retry.Status, error) {
+			var descErr error
+			// Get range descriptor (or, when spanning range, descriptors).
+			// sendAttempt below may clear them on certain errors, so we
+			// refresh (likely from the cache) on every retry.
+			desc, descNext, descErr = ds.getDescriptors(call)
+			// If getDescriptors fails, we don't fish for retryable errors.
+			if descErr != nil {
+				return retry.Break, descErr
 			}
-
-			if call.Reply != reply {
-				// This is a multi-range request. Combine the new response
-				// with the existing response.
-				call.Reply.(proto.Combinable).Combine(reply)
+			// Truncate the request to our current range, making sure not to
+			// touch it unless we have to (it is illegal to send EndKey on
+			// commands which do not operate on ranges).
+			if descNext != nil {
+				args.Header().EndKey = desc.EndKey
 			}
-			return retry.Break, err
+			return ds.sendAttempt(desc, call)
 		})
-
-		// "Untruncate" EndKey to original. We do this even on error in
-		// case the caller will retry using the same args.
+		// "Untruncate" EndKey to original.
 		args.Header().EndKey = endKey
```

Comment:
This is superfluous if `descNext` above is nil - perhaps we could run an anonymous function in the `for {` and use `defer` to do this only when needed?